src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -839,22 +834,17 @@ GetModifyConnections(Task *task, bool markCritical, bool noNewTransactions)
 													 NULL);
 
 		/*
-		 * If already in a transaction, disallow expanding set of remote
-		 * transactions. That prevents some forms of distributed deadlocks.
+		 * If we're expanding the nodes that participate in the distributed
+		 * transaction, conform to MultiShardCommitProtocol.
 		 */
-		if (noNewTransactions)
+		if (MultiShardCommitProtocol == COMMIT_PROTOCOL_2PC &&
+			InCoordinatedTransaction() &&
+			XactModificationLevel == XACT_MODIFICATION_DATA)
```

Comment:
> Also, isn't this going to trigger for transactions that only touch a single shard? 

No, that's not. We set `XACT_MODIFICATION_DATA ` after the execution is done, at the end of `ExecuteSingleModifyTask()`. Thus, it helps to understand that we've already executed another command before this. 

Also, note that we use the same approach on the master.

Some execution outputs from this PR:

```SQL

-- single statement, without any transactions
insert into test VALUES (7,7);
DEBUG:  distributed statement: INSERT INTO public.test_102016 (a, b) VALUES (7, 7)
DEBUG:  Creating router plan
DEBUG:  Plan is router executable
DEBUG:  assigned task 0 to node localhost:9702
LOG:  issuing INSERT INTO public.test_102016 (a, b) VALUES (7, 7)
DETAIL:  on server localhost:9702
INSERT 0 1


-- single statement with a tx block
postgres=# BEGIN;
BEGIN
postgres=# insert into test VALUES (7,7);
DEBUG:  distributed statement: INSERT INTO public.test_102016 (a, b) VALUES (7, 7)
DEBUG:  Creating router plan
DEBUG:  Plan is router executable
DEBUG:  assigned task 0 to node localhost:9702
LOG:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 3, '2017-08-10 11:15:11.944518+03')
DETAIL:  on server localhost:9702
LOG:  issuing INSERT INTO public.test_102016 (a, b) VALUES (7, 7)
DETAIL:  on server localhost:9702
INSERT 0 1
postgres=# COMMIT;
LOG:  issuing COMMIT
DETAIL:  on server localhost:9702
COMMIT

-- set 2pc, multiple statements that go to the same node
postgres=# SET citus.multi_shard_commit_protocol TO '2pc';
SET
postgres=# BEGIN;
BEGIN
postgres=# insert into test VALUES (7,7);
DEBUG:  distributed statement: INSERT INTO public.test_102016 (a, b) VALUES (7, 7)
DEBUG:  Creating router plan
DEBUG:  Plan is router executable
DEBUG:  assigned task 0 to node localhost:9702
LOG:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 4, '2017-08-10 11:16:19.871937+03')
DETAIL:  on server localhost:9702
LOG:  issuing INSERT INTO public.test_102016 (a, b) VALUES (7, 7)
DETAIL:  on server localhost:9702
INSERT 0 1
postgres=# insert into test VALUES (7,7);
DEBUG:  distributed statement: INSERT INTO public.test_102016 (a, b) VALUES (7, 7)
DEBUG:  Creating router plan
DEBUG:  Plan is router executable
DEBUG:  assigned task 0 to node localhost:9702
LOG:  issuing INSERT INTO public.test_102016 (a, b) VALUES (7, 7)
DETAIL:  on server localhost:9702
INSERT 0 1
postgres=# COMMIT;
LOG:  issuing COMMIT
DETAIL:  on server localhost:9702
COMMIT

-- set 2pc, multiple statements that go to the different nodes
postgres=# SET citus.multi_shard_commit_protocol TO '2pc';

postgres=# BEGIN;
BEGIN
postgres=# insert into test VALUES (7,7);
DEBUG:  distributed statement: INSERT INTO public.test_102016 (a, b) VALUES (7, 7)
DEBUG:  Creating router plan
DEBUG:  Plan is router executable
DEBUG:  assigned task 0 to node localhost:9702
LOG:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 5, '2017-08-10 11:16:49.273187+03')
DETAIL:  on server localhost:9702
LOG:  issuing INSERT INTO public.test_102016 (a, b) VALUES (7, 7)
DETAIL:  on server localhost:9702
INSERT 0 1
postgres=# insert into test VALUES (8,7);
DEBUG:  distributed statement: INSERT INTO public.test_102008 (a, b) VALUES (8, 7)
DEBUG:  Creating router plan
DEBUG:  Plan is router executable
DEBUG:  assigned task 0 to node localhost:9700
LOG:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 5, '2017-08-10 11:16:49.273187+03')
DETAIL:  on server localhost:9700
LOG:  issuing INSERT INTO public.test_102008 (a, b) VALUES (8, 7)
DETAIL:  on server localhost:9700
INSERT 0 1
postgres=# COMMIT;
LOG:  issuing PREPARE TRANSACTION 'citus_0_78762_5_0'
DETAIL:  on server localhost:9702
LOG:  issuing PREPARE TRANSACTION 'citus_0_78762_5_1'
DETAIL:  on server localhost:9700
LOG:  issuing COMMIT PREPARED 'citus_0_78762_5_0'
DETAIL:  on server localhost:9702
LOG:  issuing COMMIT PREPARED 'citus_0_78762_5_1'
DETAIL:  on server localhost:9700
COMMIT

-- and finally single statement with 2PC set

postgres=# BEGIN;
BEGIN
postgres=# insert into test VALUES (8,7);
DEBUG:  distributed statement: INSERT INTO public.test_102008 (a, b) VALUES (8, 7)
DEBUG:  Creating router plan
DEBUG:  Plan is router executable
DEBUG:  assigned task 0 to node localhost:9700
LOG:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 6, '2017-08-10 11:17:21.274619+03')
DETAIL:  on server localhost:9700
LOG:  issuing INSERT INTO public.test_102008 (a, b) VALUES (8, 7)
DETAIL:  on server localhost:9700
INSERT 0 1
postgres=# COMMIT;
LOG:  issuing COMMIT
DETAIL:  on server localhost:9700
COMMIT

```
(Sorry for so long comment 😄 )